File: app/workflows/ai_example_workflows.py
# AIDEV-NOTE: Example workflows demonstrating multi-AI provider patterns
import asyncio
from temporalio import workflow
from temporalio.workflow import execute_activity
from datetime import timedelta

from app.models.ai_models import (
    AIRequest, AIResponse, MultiAIWorkflowInput, MultiAIWorkflowResult,
    GeminiActivities, OpenAIActivities, AnthropicActivities
)


@workflow.defn(name="ai-consensus-workflow")
class AIConsensusWorkflow:
    """
    Workflow that sends the same prompt to all three AI providers
    and creates a consensus response
    """
    
    @workflow.run
    async def run(self, input: MultiAIWorkflowInput) -> MultiAIWorkflowResult:
        # AIDEV-NOTE: Execute activities in parallel for better performance
        tasks = []
        
        if "gemini" in input.providers:
            gemini_request = AIRequest(prompt=input.initial_prompt, file_path=input.file_path)
            tasks.append(
                execute_activity(
                    GeminiActivities.process_request,
                    gemini_request,
                    task_queue="gemini-ai-queue",
                    start_to_close_timeout=timedelta(minutes=5),
                )
            )
        
        if "openai" in input.providers:
            openai_request = AIRequest(prompt=input.initial_prompt, file_path=input.file_path)
            tasks.append(
                execute_activity(
                    OpenAIActivities.process_request,
                    openai_request,
                    task_queue="openai-ai-queue",
                    start_to_close_timeout=timedelta(minutes=5),
                )
            )
        
        if "anthropic" in input.providers:
            anthropic_request = AIRequest(prompt=input.initial_prompt, file_path=input.file_path)
            tasks.append(
                execute_activity(
                    AnthropicActivities.process_request,
                    anthropic_request,
                    task_queue="anthropic-ai-queue",
                    start_to_close_timeout=timedelta(minutes=5),
                )
            )
        
        # Wait for all responses
        responses = await asyncio.gather(*tasks)
        
        # Build results dictionary
        results = {}
        for i, provider in enumerate(input.providers):
            results[provider] = responses[i]
        
        # Generate consensus using Anthropic (could be any provider)
        consensus_prompt = "Based on these AI responses, create a consensus summary:\n\n"
        for provider, response in results.items():
            consensus_prompt += f"**{provider.upper()} Response:**\n{response.content}\n\n"
        consensus_prompt += "Create a balanced consensus that incorporates the best insights from all responses."
        
        consensus_response = await execute_activity(
            AnthropicActivities.process_request,
            AIRequest(prompt=consensus_prompt),
            task_queue="anthropic-ai-queue",
            start_to_close_timeout=timedelta(minutes=5),
        )
        
        return MultiAIWorkflowResult(
            results=results,
            consensus=consensus_response.content,
            analysis=f"Processed with {len(input.providers)} AI providers"
        )


@workflow.defn(name="ai-chain-workflow")
class AIChainWorkflow:
    """
    Workflow that chains AI providers: 
    Gemini analyzes -> OpenAI refines -> Anthropic validates
    """
    
    @workflow.run
    async def run(self, input: MultiAIWorkflowInput) -> MultiAIWorkflowResult:
        results = {}
        
        # Step 1: Initial analysis with Gemini
        gemini_response = await execute_activity(
            GeminiActivities.process_request,
            AIRequest(
                prompt=f"Analyze this request and provide initial insights: {input.initial_prompt}",
                file_path=input.file_path
            ),
            task_queue="gemini-ai-queue",
            start_to_close_timeout=timedelta(minutes=5),
        )
        results["gemini"] = gemini_response
        
        # Step 2: Refinement with OpenAI
        refine_prompt = f"""
        Original request: {input.initial_prompt}
        
        Initial analysis from another AI:
        {gemini_response.content}
        
        Please refine and enhance this analysis with additional insights and improvements.
        """
        
        openai_response = await execute_activity(
            OpenAIActivities.process_request,
            AIRequest(prompt=refine_prompt, file_path=input.file_path),
            task_queue="openai-ai-queue",
            start_to_close_timeout=timedelta(minutes=5),
        )
        results["openai"] = openai_response
        
        # Step 3: Validation and final polish with Anthropic
        validate_prompt = f"""
        Original request: {input.initial_prompt}
        
        Analysis progression:
        1. Initial: {gemini_response.content}
        2. Refined: {openai_response.content}
        
        Please validate the analysis, correct any errors, and provide a final polished response.
        """
        
        anthropic_response = await execute_activity(
            AnthropicActivities.process_request,
            AIRequest(prompt=validate_prompt),
            task_queue="anthropic-ai-queue",
            start_to_close_timeout=timedelta(minutes=5),
        )
        results["anthropic"] = anthropic_response
        
        return MultiAIWorkflowResult(
            results=results,
            consensus=anthropic_response.content,
            analysis="Sequential processing: Gemini → OpenAI → Anthropic"
        )


@workflow.defn(name="ai-specialist-workflow")
class AISpecialistWorkflow:
    """
    Workflow that uses different AI providers for their strengths:
    - Gemini for vision/multimodal tasks
    - OpenAI for code generation
    - Anthropic for analysis and reasoning
    """
    
    @workflow.run
    async def run(self, input: MultiAIWorkflowInput) -> MultiAIWorkflowResult:
        results = {}
        tasks = []
        
        # AIDEV-NOTE: Specialized prompts for each provider based on strengths
        
        # Gemini: Best for multimodal/vision tasks
        if input.file_path and input.file_path.lower().endswith(('.png', '.jpg', '.jpeg', '.gif')):
            gemini_prompt = f"Analyze this image in detail and {input.initial_prompt}"
        else:
            gemini_prompt = f"Provide a creative and comprehensive response to: {input.initial_prompt}"
        
        tasks.append(
            execute_activity(
                GeminiActivities.process_request,
                AIRequest(prompt=gemini_prompt, file_path=input.file_path),
                task_queue="gemini-ai-queue",
                start_to_close_timeout=timedelta(minutes=5),
            )
        )
        
        # OpenAI: Best for code and technical tasks
        openai_prompt = f"""
        Technical analysis requested: {input.initial_prompt}
        
        Please provide:
        1. Technical implementation details
        2. Code examples if applicable
        3. Best practices and considerations
        """
        
        tasks.append(
            execute_activity(
                OpenAIActivities.process_request,
                AIRequest(prompt=openai_prompt, file_path=input.file_path),
                task_queue="openai-ai-queue",
                start_to_close_timeout=timedelta(minutes=5),
            )
        )
        
        # Anthropic: Best for reasoning and analysis
        anthropic_prompt = f"""
        Analytical task: {input.initial_prompt}
        
        Please provide:
        1. Logical analysis and reasoning
        2. Potential challenges and solutions
        3. Strategic recommendations
        """
        
        tasks.append(
            execute_activity(
                AnthropicActivities.process_request,
                AIRequest(prompt=anthropic_prompt, file_path=input.file_path),
                task_queue="anthropic-ai-queue",
                start_to_close_timeout=timedelta(minutes=5),
            )
        )
        
        # Execute all tasks in parallel
        responses = await asyncio.gather(*tasks)
        
        results["gemini"] = responses[0]
        results["openai"] = responses[1]
        results["anthropic"] = responses[2]
        
        # Synthesize results
        synthesis_prompt = f"""
        Synthesize these specialized analyses into a comprehensive response:
        
        Visual/Creative Analysis (Gemini):
        {responses[0].content}
        
        Technical Analysis (OpenAI):
        {responses[1].content}
        
        Strategic Analysis (Anthropic):
        {responses[2].content}
        
        Create a unified response that leverages the strengths of each analysis.
        """
        
        final_response = await execute_activity(
            AnthropicActivities.process_request,
            AIRequest(prompt=synthesis_prompt),
            task_queue="anthropic-ai-queue",
            start_to_close_timeout=timedelta(minutes=5),
        )
        
        return MultiAIWorkflowResult(
            results=results,
            consensus=final_response.content,
            analysis="Specialized processing: Gemini (vision/creative) + OpenAI (technical) + Anthropic (analytical)"
        )

File: src/TemporalAI/Workflows/AIExampleWorkflows.cs
// AIDEV-NOTE: Example workflows demonstrating multi-AI provider patterns
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using Temporalio.Activities;
using Temporalio.Workflows;
using TemporalAI.Models;

namespace TemporalAI.Workflows
{
    /// <summary>
    /// Workflow that sends the same prompt to all three AI providers
    /// and creates a consensus response
    /// </summary>
    [Workflow("ai-consensus-workflow")]
    public class AIConsensusWorkflow
    {
        [WorkflowRun]
        public async Task<MultiAIWorkflowResult> RunAsync(MultiAIWorkflowInput input)
        {
            // AIDEV-NOTE: Execute activities in parallel for better performance
            var tasks = new List<Task<AIResponse>>();
            
            if (input.Providers.Contains("gemini"))
            {
                var geminiRequest = new AIRequest 
                { 
                    Prompt = input.InitialPrompt, 
                    FilePath = input.FilePath 
                };
                tasks.Add(Workflow.ExecuteActivityAsync<IGeminiActivities, AIResponse>(
                    a => a.ProcessRequestAsync(geminiRequest),
                    new ActivityOptions
                    {
                        TaskQueue = "gemini-ai-queue",
                        StartToCloseTimeout = TimeSpan.FromMinutes(5)
                    }
                ));
            }
            
            if (input.Providers.Contains("openai"))
            {
                var openaiRequest = new AIRequest 
                { 
                    Prompt = input.InitialPrompt, 
                    FilePath = input.FilePath 
                };
                tasks.Add(Workflow.ExecuteActivityAsync<IOpenAIActivities, AIResponse>(
                    a => a.ProcessRequestAsync(openaiRequest),
                    new ActivityOptions
                    {
                        TaskQueue = "openai-ai-queue",
                        StartToCloseTimeout = TimeSpan.FromMinutes(5)
                    }
                ));
            }
            
            if (input.Providers.Contains("anthropic"))
            {
                var anthropicRequest = new AIRequest 
                { 
                    Prompt = input.InitialPrompt, 
                    FilePath = input.FilePath 
                };
                tasks.Add(Workflow.ExecuteActivityAsync<IAnthropicActivities, AIResponse>(
                    a => a.ProcessRequestAsync(anthropicRequest),
                    new ActivityOptions
                    {
                        TaskQueue = "anthropic-ai-queue",
                        StartToCloseTimeout = TimeSpan.FromMinutes(5)
                    }
                ));
            }
            
            // Wait for all responses
            var responses = await Task.WhenAll(tasks);
            
            // Build results dictionary
            var results = new Dictionary<string, AIResponse>();
            for (int i = 0; i < input.Providers.Count; i++)
            {
                results[input.Providers[i]] = responses[i];
            }
            
            // Generate consensus using Anthropic (could be any provider)
            var consensusPrompt = "Based on these AI responses, create a consensus summary:\n\n";
            foreach (var (provider, response) in results)
            {
                consensusPrompt += $"**{provider.ToUpper()} Response:**\n{response.Content}\n\n";
            }
            consensusPrompt += "Create a balanced consensus that incorporates the best insights from all responses.";
            
            var consensusResponse = await Workflow.ExecuteActivityAsync<IAnthropicActivities, AIResponse>(
                a => a.ProcessRequestAsync(new AIRequest { Prompt = consensusPrompt }),
                new ActivityOptions
                {
                    TaskQueue = "anthropic-ai-queue",
                    StartToCloseTimeout = TimeSpan.FromMinutes(5)
                }
            );
            
            return new MultiAIWorkflowResult
            {
                Results = results,
                Consensus = consensusResponse.Content,
                Analysis = $"Processed with {input.Providers.Count} AI providers"
            };
        }
    }

    /// <summary>
    /// Workflow that chains AI providers: 
    /// Gemini analyzes -> OpenAI refines -> Anthropic validates
    /// </summary>
    [Workflow("ai-chain-workflow")]
    public class AIChainWorkflow
    {
        [WorkflowRun]
        public async Task<MultiAIWorkflowResult> RunAsync(MultiAIWorkflowInput input)
        {
            var results = new Dictionary<string, AIResponse>();
            
            // Step 1: Initial analysis with Gemini
            var geminiResponse = await Workflow.ExecuteActivityAsync<IGeminiActivities, AIResponse>(
                a => a.ProcessRequestAsync(new AIRequest
                {
                    Prompt = $"Analyze this request and provide initial insights: {input.InitialPrompt}",
                    FilePath = input.FilePath
                }),
                new ActivityOptions
                {
                    TaskQueue = "gemini-ai-queue",
                    StartToCloseTimeout = TimeSpan.FromMinutes(5)
                }
            );
            results["gemini"] = geminiResponse;
            
            // Step 2: Refinement with OpenAI
            var refinePrompt = $@"
Original request: {input.InitialPrompt}

Initial analysis from another AI:
{geminiResponse.Content}

Please refine and enhance this analysis with additional insights and improvements.";
            
            var openaiResponse = await Workflow.ExecuteActivityAsync<IOpenAIActivities, AIResponse>(
                a => a.ProcessRequestAsync(new AIRequest 
                { 
                    Prompt = refinePrompt, 
                    FilePath = input.FilePath 
                }),
                new ActivityOptions
                {
                    TaskQueue = "openai-ai-queue",
                    StartToCloseTimeout = TimeSpan.FromMinutes(5)
                }
            );
            results["openai"] = openaiResponse;
            
            // Step 3: Validation and final polish with Anthropic
            var validatePrompt = $@"
Original request: {input.InitialPrompt}

Analysis progression:
1. Initial: {geminiResponse.Content}
2. Refined: {openaiResponse.Content}

Please validate the analysis, correct any errors, and provide a final polished response.";
            
            var anthropicResponse = await Workflow.ExecuteActivityAsync<IAnthropicActivities, AIResponse>(
                a => a.ProcessRequestAsync(new AIRequest { Prompt = validatePrompt }),
                new ActivityOptions
                {
                    TaskQueue = "anthropic-ai-queue",
                    StartToCloseTimeout = TimeSpan.FromMinutes(5)
                }
            );
            results["anthropic"] = anthropicResponse;
            
            return new MultiAIWorkflowResult
            {
                Results = results,
                Consensus = anthropicResponse.Content,
                Analysis = "Sequential processing: Gemini → OpenAI → Anthropic"
            };
        }
    }

    /// <summary>
    /// Workflow that uses different AI providers for their strengths:
    /// - Gemini for vision/multimodal tasks
    /// - OpenAI for code generation
    /// - Anthropic for analysis and reasoning
    /// </summary>
    [Workflow("ai-specialist-workflow")]
    public class AISpecialistWorkflow
    {
        [WorkflowRun]
        public async Task<MultiAIWorkflowResult> RunAsync(MultiAIWorkflowInput input)
        {
            var results = new Dictionary<string, AIResponse>();
            var tasks = new List<Task<AIResponse>>();
            
            // AIDEV-NOTE: Specialized prompts for each provider based on strengths
            
            // Gemini: Best for multimodal/vision tasks
            string geminiPrompt;
            if (!string.IsNullOrEmpty(input.FilePath) && 
                (input.FilePath.EndsWith(".png", StringComparison.OrdinalIgnoreCase) ||
                 input.FilePath.EndsWith(".jpg", StringComparison.OrdinalIgnoreCase) ||
                 input.FilePath.EndsWith(".jpeg", StringComparison.OrdinalIgnoreCase) ||
                 input.FilePath.EndsWith(".gif", StringComparison.OrdinalIgnoreCase)))
            {
                geminiPrompt = $"Analyze this image in detail and {input.InitialPrompt}";
            }
            else
            {
                geminiPrompt = $"Provide a creative and comprehensive response to: {input.InitialPrompt}";
            }
            
            tasks.Add(Workflow.ExecuteActivityAsync<IGeminiActivities, AIResponse>(
                a => a.ProcessRequestAsync(new AIRequest 
                { 
                    Prompt = geminiPrompt, 
                    FilePath = input.FilePath 
                }),
                new ActivityOptions
                {
                    TaskQueue = "gemini-ai-queue",
                    StartToCloseTimeout = TimeSpan.FromMinutes(5)
                }
            ));
            
            // OpenAI: Best for code and technical tasks
            var openaiPrompt = $@"
Technical analysis requested: {input.InitialPrompt}

Please provide:
1. Technical implementation details
2. Code examples if applicable
3. Best practices and considerations";
            
            tasks.Add(Workflow.ExecuteActivityAsync<IOpenAIActivities, AIResponse>(
                a => a.ProcessRequestAsync(new AIRequest 
                { 
                    Prompt = openaiPrompt, 
                    FilePath = input.FilePath 
                }),
                new ActivityOptions
                {
                    TaskQueue = "openai-ai-queue",
                    StartToCloseTimeout = TimeSpan.FromMinutes(5)
                }
            ));
            
            // Anthropic: Best for reasoning and analysis
            var anthropicPrompt = $@"
Analytical task: {input.InitialPrompt}

Please provide:
1. Logical analysis and reasoning
2. Potential challenges and solutions
3. Strategic recommendations";
            
            tasks.Add(Workflow.ExecuteActivityAsync<IAnthropicActivities, AIResponse>(
                a => a.ProcessRequestAsync(new AIRequest 
                { 
                    Prompt = anthropicPrompt, 
                    FilePath = input.FilePath 
                }),
                new ActivityOptions
                {
                    TaskQueue = "anthropic-ai-queue",
                    StartToCloseTimeout = TimeSpan.FromMinutes(5)
                }
            ));
            
            // Execute all tasks in parallel
            var responses = await Task.WhenAll(tasks);
            
            results["gemini"] = responses[0];
            results["openai"] = responses[1];
            results["anthropic"] = responses[2];
            
            // Synthesize results
            var synthesisPrompt = $@"
Synthesize these specialized analyses into a comprehensive response:

Visual/Creative Analysis (Gemini):
{responses[0].Content}

Technical Analysis (OpenAI):
{responses[1].Content}

Strategic Analysis (Anthropic):
{responses[2].Content}

Create a unified response that leverages the strengths of each analysis.";
            
            var finalResponse = await Workflow.ExecuteActivityAsync<IAnthropicActivities, AIResponse>(
                a => a.ProcessRequestAsync(new AIRequest { Prompt = synthesisPrompt }),
                new ActivityOptions
                {
                    TaskQueue = "anthropic-ai-queue",
                    StartToCloseTimeout = TimeSpan.FromMinutes(5)
                }
            );
            
            return new MultiAIWorkflowResult
            {
                Results = results,
                Consensus = finalResponse.Content,
                Analysis = "Specialized processing: Gemini (vision/creative) + OpenAI (technical) + Anthropic (analytical)"
            };
        }
    }
}